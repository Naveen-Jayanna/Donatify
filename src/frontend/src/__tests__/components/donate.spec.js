import renderer from 'react-test-renderer';
import Donate from '../../components/donate';
import React from 'react';

describe('donateComponent', () => {
	it('renders without crashing', () => {
		const component = renderer.create(<Donate />);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('initial state set correctly', () => {
		const component = renderer.create(<Donate />);
		const expectedOutput = {
			itemName: '',
			itemQuantity: 1,
			itemDescription: '',
			itemZipCode: '',
			itemCity: {},
			itemDonorId: undefined,
			itemCategory: {},
			loading: false,
			selectedFile: null,
			imgName: ''
		};
		const actualOutput = component.getInstance().state;
		expect(expectedOutput).toEqual(actualOutput);
	});

	it('handleInput function updates state correctly - input', () => {
		const component = renderer.create(<Donate />);
		const expectedOutput = {
			itemName: 'test',
			itemQuantity: 1,
			itemDescription: '',
			itemZipCode: '',
			itemCity: {},
			itemDonorId: undefined,
			itemCategory: {},
			loading: false,
			selectedFile: null,
			imgName: ''
		};
		const input = {
			type: 'change',
			target: {
				id: 'itemName',
				value: 'test'
			}
		};
		component.getInstance().handleInput(input);
		const actualOutput = component.getInstance().state;
		expect(expectedOutput).toEqual(actualOutput);
	});

	it('handleInput function updates state correctly - dropdown', () => {
		const component = renderer.create(<Donate />);
		const expectedOutput = {
			itemName: '',
			itemQuantity: 1,
			itemDescription: '',
			itemZipCode: '',
			itemCity: {
				label: 'test',
				value: 'test'
			},
			itemDonorId: undefined,
			itemCategory: {},
			loading: false,
			selectedFile: null,
			imgName: ''
		};
		const input = {
			name: 'itemCity',
			values: {
				label: 'test',
				value: 'test'
			}
		};
		component.getInstance().handleInput(input);
		const actualOutput = component.getInstance().state;
		expect(expectedOutput).toEqual(actualOutput);
	});

	it('handleSubmit function returns false if itemName missing', (done) =>{
		const component = renderer.create(<Donate />);
		component.getInstance().setState({
			itemName: ''
		});
		component.getInstance().handleSubmit({}).then((val) => {
			expect(val).toEqual(false);
			done();
		}).catch((err) => {
			done(err);
		});
	});

	it('handleSubmit function returns false if itemDescription missing', (done) =>{
		const component = renderer.create(<Donate />);
		component.getInstance().setState({
			itemName: 'test',
			itemDescription: ''
		});
		component.getInstance().handleSubmit({}).then((val) => {
			expect(val).toEqual(false);
			done();
		}).catch((err) => {
			done(err);
		});
	});

	it('handleSubmit function returns false if itemZipCode missing', (done) =>{
		const component = renderer.create(<Donate />);
		component.getInstance().setState({
			itemName: 'test',
			itemDescription: 'test',
			itemZipCode: ''
		});
		component.getInstance().handleSubmit({}).then((val) => {
			expect(val).toEqual(false);
			done();
		}).catch((err) => {
			done(err);
		});
	});

	it('handleSubmit function returns false if itemCity missing', (done) =>{
		const component = renderer.create(<Donate />);
		window.alert = jest.fn();
		component.getInstance().setState({
			itemName: 'test',
			itemDescription: 'test',
			itemZipCode: 'test',
			itemCity: {}
		});
		component.getInstance().handleSubmit({preventDefault: () => {}}).then((val) => {
			expect(val).toEqual(false);
			expect(window.alert).toHaveBeenCalledWith('Missing value for city. Enter city for the item.');
			done();
		}).catch((err) => {
			done(err);
		});
	});

	it('handleSubmit function returns false if itemCategory missing', (done) =>{
		const component = renderer.create(<Donate />);
		window.alert = jest.fn();
		component.getInstance().setState({
			itemName: 'test',
			itemDescription: 'test',
			itemZipCode: 'test',
			itemCity: {test: 'test'}
		});
		component.getInstance().handleSubmit({preventDefault: () => {}}).then((val) => {
			expect(val).toEqual(false);
			expect(window.alert).toHaveBeenCalledWith('Missing value for category. Enter category for the item.');
			done();
		}).catch((err) => {
			done(err);
		});
	});

	it('handleSubmit function returns false if props do not exist', (done) =>{
		const component = renderer.create(<Donate />);
		component.getInstance().setState({
			itemName: 'test',
			itemDescription: 'test',
			itemZipCode: 'test',
			itemCity: {test: 'test'},
			itemCategory: {test: 'test'}
		});
		component.getInstance().handleSubmit({preventDefault: () => {}}).then((val) => {
			expect(val).toEqual(false);
			done();
		}).catch((err) => {
			done(err);
		});
	});

	it('handleSubmit function returns false if API calls fails - no error message', (done) =>{
		const props = {
			onAddItem: () => {},
			apiStatus: false
		};
		const component = renderer.create(<Donate props={props}/>);
		window.alert = jest.fn();
		component.getInstance().setState({
			itemName: 'test',
			itemDescription: 'test',
			itemZipCode: 'test',
			itemCity: {test: 'test'},
			itemCategory: {test: 'test'},
			selectedFile: 'test',
			imgName: 'test'
		});
		component.getInstance().handleSubmit({preventDefault: () => {}}).then((val) => {
			expect(val).toEqual(false);
			expect(window.alert).toHaveBeenCalledWith('Item addition could not complete. Please try again.');
			expect(component.getInstance().state.loading).toEqual(false);
			done();
		}).catch((err) => {
			done(err);
		});
	});

	it('handleSubmit function returns false if API calls fails - with error message', (done) =>{
		const props = {
			onAddItem: () => {},
			apiStatus: false,
			apiMessage: 'test'
		};
		const component = renderer.create(<Donate props={props}/>);
		window.alert = jest.fn();
		component.getInstance().setState({
			itemName: 'test',
			itemDescription: 'test',
			itemZipCode: 'test',
			itemCity: {test: 'test'},
			itemCategory: {test: 'test'},
			selectedFile: 'test',
			imgName: 'test'
		});
		component.getInstance().handleSubmit({preventDefault: () => {}}).then((val) => {
			expect(val).toEqual(false);
			expect(window.alert).toHaveBeenCalledWith('test');
			expect(component.getInstance().state.loading).toEqual(false);
			done();
		}).catch((err) => {
			done(err);
		});
	});

	it('handleSubmit function returns true if API calls succeeds', (done) =>{
		const mockFn = jest.fn();
		const props = {
			onAddItem: mockFn,
			apiStatus: true
		};
		const currentState = {
			itemName: 'test',
			itemDescription: 'test',
			itemZipCode: 'test',
			itemCity: {label: 'test', value: 'test'},
			itemCategory: {label: 'test', value: 'test'},
			selectedFile: 'test',
			imgName: 'test'
		};
		const expectedOutput = {
			itemName: 'test',
			itemDescription: 'test',
			itemDonorId: undefined,
			itemQuantity: 1,
			itemZipCode: 'test',
			itemCity: 'test',
			itemCategory: 'test',
			imgName: 'test'
		};
		const component = renderer.create(<Donate props={props}/>);
		window.alert = jest.fn();
		component.getInstance().setState(currentState);
		component.getInstance().redirectToPath = jest.fn();
		component.getInstance().handleSubmit({preventDefault: () => {}}).then((val) => {
			expect(val).toEqual(true);
			expect(component.getInstance().state.loading).toEqual(true);
			expect(mockFn).toHaveBeenCalledWith(expectedOutput);
			done();
		}).catch((err) => {
			done(err);
		});
	});

	it('redirect function updates state correctly', () => {
		const component = renderer.create(<Donate />);
		component.getInstance().redirectToPath('/');
		expect(document.location.href).toEqual('http://localhost/');
	});

	it('state updates when itemName input value changes', () => {
		const props = {
			onAddItem: () => {},
			apiStatus: true
		};
		const onChangeInput = {
			type: 'change',
			target: {
				id: 'itemName',
				value: 'test1'
			}
		};
		const component = renderer.create(<Donate props={props}/>);
		component.root.findAllByType('input')[0].props.onChange(onChangeInput);
		expect(component.getInstance().state.itemName).toEqual('test1');
	});

	it('state updates when itemDescription input value changes', () => {
		const props = {
			onAddItem: () => {},
			apiStatus: true
		};
		const onChangeInput = {
			type: 'change',
			target: {
				id: 'itemDescription',
				value: 'test1'
			}
		};
		const component = renderer.create(<Donate props={props}/>);
		component.root.findAllByType('textarea')[0].props.onChange(onChangeInput);
		expect(component.getInstance().state.itemDescription).toEqual('test1');
	});

	it('state updates when itemZipCode input value changes', () => {
		const props = {
			onAddItem: () => {},
			apiStatus: true
		};
		const onChangeInput = {
			type: 'change',
			target: {
				id: 'itemZipCode',
				value: 'test1'
			}
		};
		const component = renderer.create(<Donate props={props}/>);
		component.root.findAllByType('input')[1].props.onChange(onChangeInput);
		expect(component.getInstance().state.itemZipCode).toEqual('test1');
	});

	it('state updates when itemCity input value changes', () => {
		const props = {
			onAddItem: () => {},
			apiStatus: true
		};
		const onChangeInput = {
			type: 'change',
			target: {
				id: 'itemCity',
				value: 'test1'
			}
		};
		const component = renderer.create(<Donate props={props}/>);
		component.root.findAllByType('input')[1].props.onChange(onChangeInput);
		expect(component.getInstance().state.itemCity).toEqual('test1');
	});

	it('state updates when itemCategory input value changes', () => {
		const props = {
			onAddItem: () => {},
			apiStatus: true
		};
		const onChangeInput = {
			type: 'change',
			target: {
				id: 'itemCategory',
				value: 'test1'
			}
		};
		const component = renderer.create(<Donate props={props}/>);
		component.root.findAllByType('input')[1].props.onChange(onChangeInput);
		expect(component.getInstance().state.itemCategory).toEqual('test1');
	});

	it('renders spinner when loading is true', () => {
		const props = {
			onAddItem: () => {},
			apiStatus: true
		};
		const component = renderer.create(<Donate props={props}/>);
		component.getInstance().setState({
			loading: true
		});
		let spinner = component.root.findAllByType('span').filter((comp) => comp.props && comp.props.className === 'visually-hidden');
		expect(spinner && spinner[0] && spinner[0].props && spinner[0].props.children).toEqual('Loading...');
	});

	it('renders input when loading is false', (done) => {
		const props = {
			onAddItem: () => {},
			apiStatus: true
		};
		const component = renderer.create(<Donate props={props}/>);
		component.root.findAllByType('input')[7].props.onClick().then((val) => {
			expect(val).toEqual(false);
			done();
		}).catch((err) => {
			done(err);
		});
	});
});
