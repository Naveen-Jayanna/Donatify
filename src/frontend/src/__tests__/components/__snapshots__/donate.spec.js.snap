// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`donateComponent renders without crashing 1`] = `
<section>
  <div
    className="container"
  >
    <div
      className="signup-content"
    >
      <div
        className="signup-form"
      >
        <h2
          className="form-title"
        >
          Donate
        </h2>
        <form
          className="register-form"
          id="donate-form"
        >
          <div
            className="form-group"
          >
            <img
              alt="item name"
              src="../signup-name.png"
            />
            <input
              autoFocus={true}
              id="itemName"
              name="name"
              onChange={[Function]}
              placeholder="Item name"
              required={true}
              type="text"
              value=""
            />
          </div>
          <div
            className="form-group"
          >
            <img
              alt="item description"
              src="../item-description.png"
            />
            <textarea
              id="itemDescription"
              name="description"
              onChange={[Function]}
              placeholder="Item description"
              required={true}
              value=""
            />
          </div>
          <div
            className="form-group"
          >
            <img
              alt="item zipcode"
              src="../signup-zip.png"
            />
            <input
              id="itemZipCode"
              name="zipcode"
              onChange={[Function]}
              placeholder="Item zipcode"
              required={true}
              type="text"
              value=""
            />
          </div>
          <div
            className="form-group"
            style={
              Object {
                "overflow": "unset",
              }
            }
          >
            <img
              alt="item city"
              src="../signup-city.png"
            />
            <div
              className=" css-b62m3t-container"
              onKeyDown={[Function]}
            >
              <span
                className="css-1f43avz-a11yText-A11yText"
                id="react-select-2-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                className="css-1f43avz-a11yText-A11yText"
              />
              <div
                className=" css-1s2u09g-control"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <div
                  className=" css-dlobye-Component"
                >
                  <div
                    className=" css-14el2xx-placeholder"
                    id="react-select-2-placeholder"
                  >
                    City
                  </div>
                  <div
                    className=" css-1u4kn00-Component"
                    data-value=""
                  >
                    <input
                      aria-autocomplete="list"
                      aria-describedby="react-select-2-placeholder"
                      aria-expanded={false}
                      aria-haspopup={true}
                      autoCapitalize="none"
                      autoComplete="off"
                      autoCorrect="off"
                      className=""
                      disabled={false}
                      id="react-select-2-input"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      role="combobox"
                      spellCheck="false"
                      style={
                        Object {
                          "background": 0,
                          "border": 0,
                          "color": "inherit",
                          "font": "inherit",
                          "gridArea": "1 / 2",
                          "label": "input",
                          "margin": 0,
                          "minWidth": "2px",
                          "opacity": 1,
                          "outline": 0,
                          "padding": 0,
                          "width": "100%",
                        }
                      }
                      tabIndex={0}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  className=" css-ny0e4k-Component"
                >
                  <span
                    className=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    className=" css-tlfecz-indicatorContainer"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <svg
                      aria-hidden="true"
                      className="css-tj5bde-Svg"
                      focusable="false"
                      height={20}
                      viewBox="0 0 20 20"
                      width={20}
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <input
                name="itemCity"
                type="hidden"
                value=""
              />
            </div>
          </div>
          <div
            className="form-group"
            style={
              Object {
                "overflow": "unset",
              }
            }
          >
            <img
              alt="signup items"
              src="../signup-groceries.png"
            />
            <div
              className=" css-b62m3t-container"
              onKeyDown={[Function]}
            >
              <span
                className="css-1f43avz-a11yText-A11yText"
                id="react-select-3-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                className="css-1f43avz-a11yText-A11yText"
              />
              <div
                className=" css-1s2u09g-control"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <div
                  className=" css-dlobye-Component"
                >
                  <div
                    className=" css-14el2xx-placeholder"
                    id="react-select-3-placeholder"
                  >
                    Category
                  </div>
                  <div
                    className=" css-1u4kn00-Component"
                    data-value=""
                  >
                    <input
                      aria-autocomplete="list"
                      aria-describedby="react-select-3-placeholder"
                      aria-expanded={false}
                      aria-haspopup={true}
                      autoCapitalize="none"
                      autoComplete="off"
                      autoCorrect="off"
                      className=""
                      disabled={false}
                      id="react-select-3-input"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      role="combobox"
                      spellCheck="false"
                      style={
                        Object {
                          "background": 0,
                          "border": 0,
                          "color": "inherit",
                          "font": "inherit",
                          "gridArea": "1 / 2",
                          "label": "input",
                          "margin": 0,
                          "minWidth": "2px",
                          "opacity": 1,
                          "outline": 0,
                          "padding": 0,
                          "width": "100%",
                        }
                      }
                      tabIndex={0}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  className=" css-ny0e4k-Component"
                >
                  <span
                    className=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    className=" css-tlfecz-indicatorContainer"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <svg
                      aria-hidden="true"
                      className="css-tj5bde-Svg"
                      focusable="false"
                      height={20}
                      viewBox="0 0 20 20"
                      width={20}
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <input
                name="itemCategory"
                type="hidden"
                value=""
              />
            </div>
          </div>
          <div>
            <input
              onChange={[Function]}
              type="file"
            />
          </div>
          <div
            className="form-group form-button"
          >
            <input
              className="form-submit"
              id="donate"
              name="donate"
              onClick={[Function]}
              type="submit"
              value="Donate"
            />
          </div>
        </form>
      </div>
      <div
        className="signup-image"
      >
        <figure>
          <img
            alt="donate"
            src="../donate-image.jpg"
          />
        </figure>
      </div>
    </div>
  </div>
</section>
`;
